src/backend/utils/activity/memtrack.c

```diff
@@ -0,0 +1,185 @@
+/*-------------------------------------------------------------------------
+ *
+ * memtrack.c
+ *	  track and manage memory usage by the PostgreSQL server.
+ *
+ * Portions Copyright (c) 1996-2023, PostgreSQL Global Development Group
+ * Portions Copyright (c) 1994, Regents of the University of California
+ *
+ *
+ * IDENTIFICATION
+ *	  src/backend/utils/activity/memtrack.c
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include <unistd.h>
+
+#include "postgres.h"
+#include "miscadmin.h"
+#include "utils/backend_status.h"
+#include "utils/memtrack.h"
+#include "storage/proc.h"
+
+/* Forward references */
+static inline bool atomic_add_with_limit(pg_atomic_uint64 *sum, uint64 add, uint64 limit);
```

Comment:
Thought we agreed to move this into the atomics header file?

Reply:
I'm all in favor of putting it in atomics.h, if we can introduce just that one routine. 

If we need to introduce a complete set of variations like PG has for atomic add, we would need 8 new routines
 - pg_atomic_fetch_add_limit_u64, pg_atomic_add_fetch_limit_u64, 
 - pg_atomic_fetch_sub_limit_u64, pg_atomic_sub_fetch_limit_u64
 -  pg_atomic_fetch_add_limit_u32, pg_atomic_add_fetch_limit_u32, 
 - pg_atomic_fetch_sub_limit_u32, pg_atomic_sub_fetch_limit_u32
 along with a new set of unit tests.

In the latter case, I'd rather hold off moving it to atomics.h.  Your thoughts?

Reply:
Let's start with just introducing the one routine and see what folks say.